internal/extproc/models_processor.go

```diff
@@ -0,0 +1,97 @@
+package extproc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log/slog"
+	"time"
+
+	corev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	extprocv3 "github.com/envoyproxy/go-control-plane/envoy/service/ext_proc/v3"
+	typev3 "github.com/envoyproxy/go-control-plane/envoy/type/v3"
+	"google.golang.org/grpc/codes"
+
+	"github.com/envoyproxy/ai-gateway/internal/apischema/openai"
+)
+
+// ModelsProcessor handles the processing of requests to the `/v1/models` endpoint.
+// This processor returns an immediate response with the list of models that are declared in the filter
+// configuration.
+// Since it returns an immediate response after processing the headers, the rest of the methods of the
+// ProcessorIface are not implemented. Those should never be called.
+type ModelsProcessor struct {
+	logger *slog.Logger
+	models openai.ModelList
+}
+
+var _ ProcessorIface = (*ModelsProcessor)(nil)
+
+// NewModelsProcessor creates a new processor that returns the list of declared models
+func NewModelsProcessor(config *processorConfig, logger *slog.Logger) ProcessorIface {
+	models := openai.ModelList{
+		Object: "list",
+		Data:   make([]openai.Model, 0, len(config.declaredModels)),
+	}
+	for _, m := range config.declaredModels {
+		models.Data = append(models.Data, openai.Model{
+			ID:      m,
+			Object:  "model",
+			OwnedBy: "Envoy AI Gateway",              // TODO(nacx): make this configurable when we need more flexibility
+			Created: openai.JSONUNIXTime(time.Now()), // TODO(nacx): does this really matter here?
+		})
+	}
+	return &ModelsProcessor{logger: logger, models: models}
+}
+
+// ProcessRequestHeaders implements [ProcessorIface.ProcessRequestHeaders].
+func (m *ModelsProcessor) ProcessRequestHeaders(_ context.Context, _ *corev3.HeaderMap) (*extprocv3.ProcessingResponse, error) {
+	m.logger.Info("Serving list of declared models")
+
+	body, err := json.Marshal(m.models)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal body: %w", err)
+	}
```

Comment:
nit: i think we could marshal during the config setup and simply have []byte in the ModelsProcessor struct vs create the byte slice each time?

Reply:
I though about that, but we're instantiating the processor once per-request, and this will be executed only once per request anyway.
If we want to save computation probably we should do this in the LoadConfig method, after loading the config, so that's available when the processors are instantiated. I didn't want to to this, though, because it couples some concrete processor behaviour (the models processor) to the Server's load config.

I also think this is a negligible overhead when returning the immediate response.

WDYT?

Reply:
yeah let's leave it as is